crates/matrix-sdk/src/event_cache/room/mod.rs

```diff
@@ -55,6 +64,62 @@ impl fmt::Debug for RoomEventCache {
     }
 }
 
+/// Thin wrapper for a room event cache listener, so as to trigger side-effects
+/// when all listeners are gone.
+#[allow(missing_debug_implementations)]
+pub struct RoomEventCacheListener {
+    /// Underlying receiver of the room event cache's updates.
+    recv: Receiver<RoomEventCacheUpdate>,
+
+    /// To which room are we listening?
+    room_id: OwnedRoomId,
+
+    /// Sender to the auto-shrink channel.
+    auto_shrink_sender: mpsc::Sender<AutoShrinkChannelPayload>,
+
+    /// Shared instance of the auto-shrinker.
+    listener_count: Arc<AtomicUsize>,
+}
+
+impl Drop for RoomEventCacheListener {
+    fn drop(&mut self) {
+        let previous_listener_count = self.listener_count.fetch_sub(1, Ordering::SeqCst);
+        if previous_listener_count == 1 {
+            // We were the last instance of the listener; let the auto-shrinker know by
+            // notifying it of our room id.
+
+            let mut room_id = self.room_id.clone();
+
+            // Try to send without waiting for channel capacity, and restart in a spin-loop
+            // if it failed; the channel shouldn't be super busy in general, so
+            // this should resolve quickly enough.
+
+            while let Err(err) = self.auto_shrink_sender.try_send(room_id) {
```

Comment:
Hmm, this loop worries me a lot. Can we panic if it executes too many times? The idea is to avoid having an infinite loop (in practise it should not happen, but better being safe than sorry).

Reply:
ok, will add a large guard, like 1024 attempts